proposals/lip-0045.md

```diff
@@ -984,164 +839,223 @@ The Interoperability module contains the following commands:
-        return False
+def send(sendingAddress: Address, 
+    moduleID: ModuleID, 
+    crossChainCommandID: CrossChainCommandID, 
+    receivingChainID: ChainID, 
+    fee: uint64, 
+    status: uint32, 
+    params: object) -> None:
+    # Processing on the mainchain
+    if ownChainAccount.ID == MAINCHAIN_ID:
+        # Chain has to exist on mainchain
+        if chainAccount(receivingChainID) does not exists:
+            raise Exception('Receiving chain does not exist.')
+        partnerChainID = receivingChainID
+    # Processing on a sidechain
+    else:
+        # Check for direct channel
+        if chainAccount(receivingChainID) does not exists:
+            partnerChainID = mainchainID
+        else:
+            partnerChainID = receivingChainID
 
-    return sendInternal(
-        EMPTY_FEE_ADDRESS,
-        ccm.moduleID,
-        ccm.crossChainCommandID,
-        ccm.sendingChainID,
-        0,
-        errorStatus,
-        ccm.params
+    # Chain must be live; This checks is always on the receivingChainID
+    if not isLive(receivingChainID):
+        raise Exception('Receiving chain is not live.')
+    # Chain status must be active
+    if not chainAccount(partnerChainID).status == CHAIN_STATUS_ACTIVE:
+        raise Exception('Receiving chain is not active.')
+
+    # Create cross-chain message
+    paramsSchema = JSON schema associated with (moduleID, crossChainCommandID)
+    ccm = {
+        "nonce": ownChainAccount.nonce,
+        "moduleID": moduleID,
+        "crossChainCommandID": crossChainCommandID,
+        "sendingChainID": ownChainAccount.ID,
+        "receivingChainID": receivingChainID,
+        "fee": fee,
+        "status": status,
+        "params": encode(paramsSchema, params)
+    }
+
+    # Validate ccm size
+    validateFormat(ccm)
+
+    # Pay message fee
+    Token.payMessageFee(sendingAddress, ccm)
+
+    addToOutbox(partnerChainID, ccm)
+    ownChainAccount.nonce += 1
+
+    # Emit CCM processed event.
+    ccmID = SHA-256(encode(crossChainMessageSchema, ccm))
+
+    emitEvent(
+        moduleID = MODULE_ID_INTEROPERABILITY,
+        typeID = TYPE_ID_CCM_PROCESSED,
+        data = {"ccmID": ccmID, "status": CCM_PROCESSED_STATUS_SENT},
+        topics = [ccm.sendingChainID, ccm.receivingChainID, ccmID]
     )
 ```
 
 #### terminateChain
 
-For the function description, parameters, and return value see the corresponding internal function `terminateChainInternal`.
-
-##### Execution
+The `terminateChain` function terminates a chain account.
 
 ```python
-terminateChain(chainID):
+def terminateChain(chainID: ChainID) -> None:
+    # Chain was already terminated, do nothing
     if terminatedStateAccount(chainID) exists:
-        return False
+        return
 
     if chainID == ownChainAccount.ID:
-        return False
+        raise Exception('Chain cannot terminate itself.')
 
     # Processing on a sidechain
     if ownChainAccount.ID != MAINCHAIN_ID:
```

Comment:
This check is not necessary, right? As the Lisk Mainchain can already not terminate itself.

Reply:
This function is an API tho.

Reply:
All good, I was just confused.